### PR TITLE
Adds support for kNN/ANN searches within Elasticsearch

### PR DESCRIPTION
**Pull request by andyHa**

### CODE REVIEW

File: src/main/java/sirius/db/es/types/DenseVector.java [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+ * neighbor searches during retrieval. Using {@link sirius.db.es.ElasticQuery#knn(NearestNeighborsSearch)} an
+ * approximate nearest neighbor search can be performed which is implemented using rather efficient HNSW graphs.
+ */
+public class DenseVector {
+
+    /**
+     * Specified the similarity function to apply.
+     */
+    public enum Similarity {
+        COSINE("cosine"), EUCLIDEAN("l2_norm"), INNER_PRODUCT("dot_product");
```
Comment (sabieber) [inline src/main/java/sirius/db/es/types/DenseVector.java:30]: **Optional:** Maybe shortly describe the different Valses

File: src/main/java/sirius/db/es/types/DenseVectorProperty.java [hunk 2]
```diff
@@ -0,0 +1,151 @@
...
+        description.put("type", "dense_vector");
+        description.put("dims", referenceVector.dimensions);
+        description.put("index", referenceVector.indexed);
+        description.put("similarity", referenceVector.similarity.getEsName());
+    }
+
+    protected DenseVector getDenseVector(Object entity) {
+        try {
+            return (DenseVector) super.getValueFromField(entity);
+        } catch (Exception e) {
```
Comment (sabieber) [inline src/main/java/sirius/db/es/types/DenseVectorProperty.java:99]: **Optional:** Rename ˋeˋ to ˋexceptionˋ

**Review by sabieber — APPROVED** [review]

File: src/main/java/sirius/db/es/types/DenseVectorProperty.java [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+            DenseVectorProperty denseVectorProperty = new DenseVectorProperty(descriptor, accessPath, field);
+
+            try {
+                if (field.get(accessPath.apply(descriptor.getType().getDeclaredConstructor().newInstance())) == null) {
+                    Mixing.LOG.WARN("Field %s in %s is a DenseVector. Such fields should be initialized as it"
+                                    + " should never be null!", field.getName(), field.getDeclaringClass().getName());
+                }
+
+                if (Modifier.isFinal(field.getModifiers())) {
+                    Mixing.LOG.WARN("Field %s in %s is a DenseVector and should be final!",
```
Comment (fhaScireum) [inline src/main/java/sirius/db/es/types/DenseVectorProperty.java:58]: Warning message does not match if-Statement. Shouldn't the boolean expression inside the if statement be negated?

File: src/main/java/sirius/db/es/types/DenseVector.java (hunk 1 above)
Comment (jakobvogel) [reply src/main/java/sirius/db/es/types/DenseVector.java:30]: …and is it good to user other names in the enum than in ES lingo?

File: src/main/java/sirius/db/es/types/DenseVectorProperty.java [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+                                    field.getName(),
+                                    field.getDeclaringClass().getName());
+                }
+            } catch (Exception e) {
```
Comment (jakobvogel) [inline src/main/java/sirius/db/es/types/DenseVectorProperty.java:62]: Another place where @sabieber's comment about variable names applies.

**Review by jakobvogel — APPROVED** [review]

File: src/main/java/sirius/db/es/types/DenseVector.java (hunk 1 above)
Comment (andyHa) [reply src/main/java/sirius/db/es/types/DenseVector.java:30]: > …and is it good to user other names in the enum than in ES lingo?

you might want to discuss that with copilot - I'm just the guy that commited.... ;-)

**Review by mkeckmkeck — APPROVED** [review]

File: src/main/java/sirius/db/es/types/DenseVector.java [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+    /**
+     * Specified the similarity function to apply.
+     */
+    public enum Similarity {
+
+        /**
+         * Cosine similarity.
+         * <p>
+         * This is the default similarity function used by Elasticsearch and also recommended for vectors generated
+         * by most embedding algorithms.However, if normalization is possible, the {@link #DOT_PRODUCT} is preferred.
```
Comment (sabieber) [inline src/main/java/sirius/db/es/types/DenseVector.java:35]:
```suggestion
         * by most embedding algorithms. However, if normalization is possible, the {@link #DOT_PRODUCT} is preferred.
```

File: src/main/java/sirius/db/es/types/DenseVector.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+        /**
+         * Represents the L-squared distance, better known as Euclidean distance.
+         */
+        L2_NORM("l2_norm"),
+
+        /**
+         * Represents the dot product of the vectors.
+         * <p>
+         * This the vectors are normalized to unit length, this is equivalent to the cosine similarity but mucht more
```
Comment (sabieber) [inline src/main/java/sirius/db/es/types/DenseVector.java:47]:
```suggestion
         * If the vectors are normalized to unit length, this is equivalent to the cosine similarity but much more
```

**Review by sabieber — APPROVED** [review]

**Review by fhaScireum — APPROVED** [review]

**Pull request merged**